### ISSUE TITLE
Karma runs tests before the preprocessor is done

### ISSUE THREAD

**Issue by jaramir** [ryanclark/karma-webpack#81]
Hello,

I'd like to run tests in IntelliJ using karma-intellij. 

IntelliJ starts a Karma server and uses HTTP POST to /run to run tests.

The problem I see is that when I change a file and run tests (shift-f10) the webpack preprocessor is still running when the tests are run so I get the results from the previous version of the file.

If I run the tests again I will get the new results (once webpack preprocessing is done).

If I save the file, wait and than run the tests it works correctly but it is an annoying workflow.

Thanks for the help!

**Comment by leodsgn** [comment]
I'm having this issue, as well.

follows my karma config:

```js
// Karma configuration
// Generated on Wed Oct 12 2016 15:25:38 GMT+0000 (UTC)
var webpack = require("webpack");
var webpackConfig = require("./webpack/test");

module.exports = function (config) {
    config.set({

        // base path that will be used to resolve all patterns (eg. files, exclude)
        basePath: '',


        // frameworks to use
        // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['mocha', 'chai', 'sinon'],


        // list of files / patterns to load in the browser
        files: [
            // all files ending in "_test"
            // {pattern: 'tests/**/*.spec.ts', watched: false}
            'tests/index.spec.ts'
            // each file acts as entry point for the webpack configuration
        ],


        // list of files to exclude
        exclude: [],


        // preprocess matching files before serving them to the browser
        // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {
            'tests/index.spec.ts': ['webpack']
        },

        webpack: {
            debug: true,
            devtool: 'eval-source-map', //'inline-source-map', - inline-source-map doesn't work at present
            module: webpackConfig.module,
            resolve: webpackConfig.resolve
        },

        webpackMiddleware: {
            quiet: true,
            stats: {
                colors: true
            }
        },

        // reporter options
        mochaReporter: {
            colors: {
                success: 'bgGreen',
                info: 'cyan',
                warning: 'bgBlue',
                error: 'bgRed'
            }
        },

        // test results reporter to use
        // possible values: 'dots', 'progress'
        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
        reporters: ['progress'],


        // web server port
        port: 9876,


        // enable / disable colors in the output (reporters and logs)
        colors: true,


        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_DEBUG,


        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: true,


        // start these browsers
        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
        browsers: ['PhantomJS'],

        plugins: ['karma-webpack', 'karma-mocha', 'karma-mocha-reporter', 'karma-chai', 'karma-sinon', 'karma-sourcemap-loader', 'karma-phantomjs-launcher'],


        // Continuous Integration mode
        // if true, Karma captures browsers, runs the tests and exits
        singleRun: false,

        // Concurrency level
        // how many browser should be started simultaneous
        concurrency: Infinity,

    })
};

```

And my webpack config:

```js
module.exports = {
    entry: './tests/index.spec.ts',
    output: {
        filename: 'dist/test.bundle.js',
        publicPath: 'dist'
    },
    resolve: {
        extensions: ['.ts', '.js', '.tsx', '.jsx', '']
    },
    module: {
        loaders: [
            {
                test: /\.tsx?$/,
                exclude: /node_modules/,
                loader: 'ts-loader'
            }
        ]
    },
    watch: true
};

```

**Comment by d3viant0ne** [comment]
@jaramir @leodsgn - If I can get a minimal repo where this issue is reproducible, i'll get it into the v3 milestone and see that it's addressed. 

Something like this - https://github.com/Silviu-Marian/karma-webpack-pr69 would be extremely helpful.

**Comment by devoto13** [comment]
I have tried to make a minimal repro for this issue in https://github.com/devoto13/karma-issue-81 and so far it looks like it's not issue of `karma-webpack`, but rather `karma-intellij` messing up something in the Karma config, which breaks `karma-webpack` from waiting for compilation. See README for current findings. 

I've also found open issue in `karma-intellij` project - https://github.com/karma-runner/karma-intellij/issues/35.

**Comment by devoto13** [comment]
Disregard my previous comment, linked repro is not reliable repro. But it was useful in investigation.

So the difference between triggering tests run by `$ karma run` and `karma-intelij` plugin is that later one sends `refresh: false` when asking Karma server to rerun tests. Same behavior can be achieved by running using `$ karma run --no-refresh`.

This flag leads to Karma server not re-fetching files and therefore `webpackBlocker` middleware not being called at all (since there is no HTTP requests for the new files sent) and therefore not blocking test run until compilation is finished. 

I see three possible solutions here:

1. `karma-intellij` should send `refresh: true` as pointed out in https://github.com/karma-runner/karma-intellij/issues/35
2. `karma-webpack` should [refresh files](https://github.com/webpack-contrib/karma-webpack/blob/master/src/karma-webpack.js#L156) not on successful compilation, but straigt after [compilation was invalidated](https://github.com/webpack-contrib/karma-webpack/blob/master/src/karma-webpack.js#L88), so Karma server will be forced to refresh files and wait in the `webpackBlocker` middleware. Not sure how it will work if compilation failed though...
3. `webpackBlocker` should block not file request, but the test run itself. If it's even possible.

What do you think @d3viant0ne, @segrey?

**Comment by segrey** [comment]
@devoto13 Thanks for the investigation. karma-intellij will pass `refresh: true`. Even though it will slow down a test run a bit, it's worth doing.

**Comment by devoto13** [comment]
@segrey Great. Thanks for the quick fix!

**Comment by devoto13** [comment]
WebStorm 2017.2.1 contains the fix and works well for me. See [release notes](https://blog.jetbrains.com/webstorm/2017/08/webstorm-2017-2-1/). I think change should be propagated to other JetBrains IDEs soon (if not already).

This issue can probably be closed now.

**Comment by codymikol** [comment]
Closing due to the age of this ticket, if you're having problems with the current build, please open a new issue and I'll help as best I can!
